java-squid/src/main/java/org/sonar/java/se/SymbolicValueRelation.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class SymbolicValueRelation {
+
+  protected final SymbolicValue v1;
+  protected final SymbolicValue v2;
+
+  protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
+    this.v1 = v1;
+    this.v2 = v2;
+  }
+
+  public Boolean impliedBy(List<SymbolicValueRelation> knownRelations) {
+    for (SymbolicValueRelation relation : knownRelations) {
+      Boolean result = isImpliedBy(relation);
+      if (result != null) {
+        return result;
+      }
+    }
+    for (int i = 0; i < knownRelations.size(); i++) {
+      SymbolicValueRelation relation = knownRelations.get(i);
+      SymbolicValueRelation combined = underHypothesis(relation);
+      if (combined != null) {
+        final List<SymbolicValueRelation> relations = new ArrayList<>(knownRelations);
+        relations.remove(i);
+        Boolean result = combined.impliedBy(relations);
+        if (result != null) {
+          return result;
+        }
+      }
+    }
+    return null; // NOSONAR
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Is //NOSONAR used to exclude false-positive or to hide real quality flaw ? [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ANoSonar)